File: barbican/tree/file/find.py
from typing import Tuple

from tree.branch.branch import Branch


def _get_token(route: str) -> Tuple[str, str]:
    end = route.find('/')
    if end == -1:
        return route, ''

    token = route[:end]
    new_route = route[(end + 1):]

    return (token, new_route)


def find(route: str, branch: Branch):
    token, new_route = _get_token(route)

    if not new_route:
        if token not in branch.headers:
            return None

        return branch.headers[token]

    elif token == '.':
        return find(new_route, branch)

    elif token == '..':
        if not token.parent:
            return None

        return find(new_route, branch.parent)

    else:
        if token not in branch.branches:
            return None

        return find(new_route, branch.branches[token])

File: source/tree/branch/find.go
package branch

import (
    "fmt"
    "errors"
    "strings"

    "github.com/inigo-selwood/barbican/tree/asset/header"
)

func trim(route string) (string, string) {
    endIndex := strings.Index(route, "/")
    if endIndex == -1 {
        return route, ""
    }

    token := route[:endIndex]
    newRoute := route[(endIndex + 1):]

    return token, newRoute
}

func Find(route string, context *Branch) (*header.Header, error) {
    if context == nil {
        return nil, errors.New("nil context")
    }

    token, newRoute := trim(route)

    // If the last token has been extracted, try to find the header in the
    // current branch
    if newRoute == "" {
        header, headerFound := context.Headers[token]
        if headerFound == false {
            error := fmt.Errorf("no such header '%s' in '%s'",
                    token,
                    context.Name)
            return nil, error
        }

        return header, nil

    // Handle arent-routing tokens
    } else if token == ".." {
        return Find(newRoute, context.Parent)

    // Handle self-routing tokens
    } else if token == "." {
        return Find(newRoute, context)
    }

    // Otherwise, try to delegate to a sub-branch
    newContext, contextFound := context.Branches[token]
    if contextFound == false {
        error := fmt.Errorf("no such branch '%s' in '%s'",
                token,
                context.Name)
        return nil, error
    }

    return Find(newRoute, newContext)
}
